### PR TITLE
fix(python): add support for pyarrow 13+

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description

I build on top of the branch of @wjones127 https://github.com/delta-io/delta-rs/pull/1602. In pyarrow v13+ the ParquetWriter by default uses the `compliant_nested_types = True` (see related PR: https://github.com/apache/arrow/pull/35146/files)and the docs: https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html).

In arrow/parquet-rs it fails when it compares schemas because it expected the old non-compliant ones. For now we can have pyarrow 13+ supported by disabling it or updating the file options provided by a user.

# Related Issue(s)

- Closes https://github.com/delta-io/delta-rs/issues/1744

# Documentation

### CODE REVIEW

File: python/deltalake/table.py [hunk 1]
```diff
@@ -1435,3 +1435,18 @@ def z_order(
...
+    """
+    Cast a batch to a schema, if it is already considered equal.
+
+    This is mostly for mapping things like list field names, which arrow-rs
+    checks when looking at schema equality, but pyarrow does not.
+    """
+    if batch.schema == schema:
+        return pyarrow.Table.from_batches([batch]).cast(schema).to_batches()[0]
+    else:
+        return batch
```
Comment (wjones127) [inline python/deltalake/table.py:1452]: @ion-elgreco If we take this out, do the tests still pass?

Comment (ion-elgreco) [reply python/deltalake/table.py:1452]: Yes, everything passes still. Shall I take it out? 

Or would we require it later once we enable the compliant types again.

Comment (wjones127) [reply python/deltalake/table.py]: Ideally I think we shouldn’t need it. This was an earlier attempt at fixing the compliant types thing

Comment (ion-elgreco) [reply python/deltalake/table.py]: We should have backwards compatibility during reads and writes when we use the compliant types, I think spark-delta also doesn't write these compliant types but I may be wrong here

**Review by wjones127 — APPROVED** [review]

**Comment by wjones127** [comment]
Thanks for finishing this up!

**Pull request merged**